src/soc/intel/baytrail/txei.c

```diff
@@ -0,0 +1,1039 @@
+{
+	return read32p(get_txei_bar1(dev) + offset);
+}
+
+static void write_bar(pci_devfn_t dev, uint32_t offset, uint32_t val)
+{
+	return write32p(get_txei_bar(dev) + offset, val);
+}
+
+static void write_bar1(pci_devfn_t dev, uint32_t offset, uint32_t val)
+{
+	return write32p(get_txei_bar1(dev) + offset, val);
+}
+
+static bool txe_ipc_ready(void)
+{
+	uint32_t ipc_input_sts;
+	ipc_input_sts = read_bar(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_SEC_IPC_INPUT_STS);
+	return !!(ipc_input_sts & 1);
+}
+
+static int wait_heci_ipc_ready(void)
+{
+	struct stopwatch sw;
+
+	stopwatch_init_usecs_expire(&sw, HECI_READY_TIMEOUT);
+	while (!txe_ipc_ready()) {
+		udelay(HECI_WAIT_DELAY);
+		if (stopwatch_expired(&sw))
+			return 0;
+	}
+
+	return 1;
+}
+
+static bool check_heci_reset(void)
+{
+	return ((read_bar1(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_SEC_IPC_READINESS) & IPC_READINESS_SEC_RDY) == 0);
+}
+
+static uint32_t read_slot(uint32_t idx)
+{
+	return read_bar1(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_IPC_OUTPUT_PAYLOAD + idx * SLOT_SIZE);
+}
+
+static void write_slot(uint32_t idx, uint32_t val)
+{
+	write_bar(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_IPC_INPUT_PAYLOAD + idx * SLOT_SIZE, val);
+}
+
+static void set_ipc_input_doorbell(void)
+{
+	uint32_t ipc_input_doorbell;
+	ipc_input_doorbell = read_bar(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_SEC_IPC_INPUT_DRBELL);
+	write_bar(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_SEC_IPC_INPUT_DRBELL, ipc_input_doorbell | 1);
+}
+
+static int ipc_output_ready(void)
+{
+	uint32_t ipc_output_sts;
+	ipc_output_sts = read_bar1(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_S_SEC_IPC_OUTPUT_STS);
+	return !!(ipc_output_sts & 1);
+}
+
+static void set_ipc_output_ready(void)
+{
+	uint32_t ipc_output_sts;
+	ipc_output_sts = read_bar1(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_S_SEC_IPC_OUTPUT_STS);
+	write_bar1(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_MMIO_SEC_IPC_OUTPUT_STS, ipc_output_sts | 1);
+}
+
+static int wait_read_slots(void)
+{
+	struct stopwatch sw;
+
+	stopwatch_init_usecs_expire(&sw, HECI_READ_TIMEOUT);
+	while (true) {
+		udelay(HECI_WAIT_DELAY);
+		if (stopwatch_expired(&sw)) {
+			printk(BIOS_ERR, "HECI: timed out reading answer!\n");
+			return 0;
+		}
+
+		if (!ipc_output_ready())
+			break;
+	}
+	return 1;
+}
+
+
+static size_t hdr_get_length(uint32_t hdr)
+{
+	return (hdr & TXEI_HDR_LENGTH) >> TXEI_HDR_LENGTH_START;
+}
+
+/* get number of full 4-byte slots */
+static size_t bytes_to_slots(size_t bytes)
+{
+	return (bytes + (SLOT_SIZE - 1) + SLOT_SIZE) / SLOT_SIZE;
+}
+
+static int
+send_one_message(uint32_t hdr, const void *buff)
+{
+	size_t pend_len, pend_slots, i;
+	const uint32_t *p = buff;
+
+	if (!wait_heci_ipc_ready())
+		return 0;
+
+	/* First, write header */
+	write_slot(0, hdr);
+
+	pend_len = hdr_get_length(hdr);
+	pend_slots = bytes_to_slots(pend_len);
+
+	/* Write the body in whole slots */
+	for (i = 1; i < pend_slots; i++, *p++) {
+		write_slot(i, *p);
```

Comment:
It may often be out of bound read, depending on the message length. If it is not divisible by 4, then some additional bytes may be sent at the end of message. However, ME/TXE should not care, because the real length of the message is in the message header. But we still must write 32bit slots... Of course we may always check for the message length, reallocate and copy it to a new buffer that is divisible by 4 and appended with zeros if necessary. However, the original source of CSE HECI `src/soc/intel/common/block/cse/cse.c` does not do it.

Reply:
https://github.com/Dasharo/coreboot/commit/6a4524e4130d3cf460d9aa66859d1f9ff2af7aa2

This is not exactly what you wanted. Although like I mentioned above, it is based on the CSE HECI from common blocks, which does the same. And TXE/CSE should not care about bytes after the message size.